Update windows runtime

mlaventure opened the pull request:
This is a cleanup/update of the windows runtime

It works with manual testing on my side, but the following is left for me to do:
- [x] Add support for the servicing container 
- [x] Remove all debug code
- [x] Ensure/Fix the unit test so they run on Windows
- [x] Look if we can use http://www.appveyor.com for CI on windows

-- 

## Issues

- This branch contains 2 fix for go-winio
- There's a fix for #1105 , at least it does the trick on my windows vm (it's the last commit)

darrenstahlmsft commented:
Working for me!

Though currently if the CreateProcess call fails the client seems to hang. Haven't looked into exactly why, but stopping the container manually doesn't unfreeze it, potentially pipes not getting closed?

mlaventure commented:
@darrenstahlmsft Thanks for testing it!

The pipes are indeed not closed by the code at the moment, I've updated the code to do so. Although, since we returned an error, the client should probably not wait for them (but we would still have leaked the pipe fds).

I've updated the code to take care of the case where  there an error when creating the process, i also added a safety call to the Close routine of the io when the task is deleted (mostly needed for the io of the init process if it is never started, but it doesn't hurt for the other processes).

I'll work on updating the test code today and fixing the bugs that I'll probably find while doing so :)

darrenstahlmsft commented:
@mlaventure Great!

How stable do you think this version of the code is? I've got a lot of work to get the LCOW runtime working and it shares a lot of common code here. To avoid merge conflicts, I'm planning on building on top of this PR. Should I hold off a bit?

mlaventure commented:
@darrenstahlmsft I haven't touch it much at the moment, I'm going through the test suite, but a few are deadlocking at the moment and I'm trying to figure out why. I'm hoping to be all done by Friday EOD (Paris Time)

mlaventure commented:
@darrenstahlmsft you should be able to use this branch now despite the 2 issues I added in the original post.

I'll work on appveyor tomorrow

darrenstahlmsft commented:
FYI I think @johnstep was also looking at getting Windows CI up and running.

Thanks! I'll start partying on this branch 👍

mlaventure commented:
Ok, so I got the CI working on `appveyor`.

However, it's failing `make coverage` on the Diff unit tests atm (cc @stevvooe @dmcgowan)

Also, someone need to give `appveyor` access to containerd (cc @crosbymichael?).

I have it activated on my fork, you can see the last result here: https://ci.appveyor.com/project/mlaventure/containerd/build/24

Unfortunately, I'll be in PTO starting in a few mins, so I won't have the time to look into the issue with diff, but any maintainers should be able to push on this branch with the bonus of any push automatically starting a build on appveyor.

I really do think this needs to be merged to avoid breaking windows with future change. But I'm conscious it's a lot of changes, I've tried to separate the not so windows changes into their own commit to make it easier.

As I said, I'll be in PTO, so feel free to add commit / make changes to get this merged :)

dmcgowan commented on cmd/ctr/.#info.go:
```diff
@@ -0,0 +1 @@
+mlaventure@tora.9070:1499667430
```
Accidental include

mlaventure replied:
Thanks, removed it!

dmcgowan commented on client_test.go:
```diff
@@ -94,14 +94,22 @@ func TestMain(m *testing.M) {
-		cmd.Wait()
-		fmt.Fprintf(os.Stderr, "%s: %s", err, buf.String())
+	if runtime.GOOS != "windows" { // TODO: remove once pull is supported on windows
+		if _, err = client.Pull(ctx, testImage, WithPullUnpack); err != nil {
```
We can correct this later but Windows should be able to test the fetch portion by not including `WithPullUnpack`

dmcgowan commented:
I agree we should get this merged so I can focus on getting the tests fixed and updated for windows, LGTM

stevvooe commented:
LGTM

stevvooe commented:
> However, it's failing make coverage on the Diff unit tests atm (cc @stevvooe @dmcgowan)

We can burn coverage, as far as I'm concerned. The only thing it is useful for is generating CO2 and creating notification in github after build finishes from the comment.

dmcgowan commented on vendor/github.com/Microsoft/go-winio/file.go:
```diff
@@ -69,6 +69,7 @@ func initIo() {
 // win32File implements Reader, Writer, and Closer on a Win32 handle without blocking in a syscall.
 // It takes ownership of this handle and will close it if it is garbage collected.
 type win32File struct {
+	sync.Mutex
```
Something is off about this update, I don't see these changes in `v0.4.3`

dmcgowan replied:
For reference https://github.com/Microsoft/go-winio/commits/v0.4.3/file.go

dmcgowan replied:
I see what you did there, https://github.com/Microsoft/go-winio/pull/61

In the future, can you put a comment in vendor.conf and point it to your fork with the correct hash.

mlaventure replied:
There was actually 2 PR integrated in this one.

Sorry, I should have made it into a proper fork, I'll remember it next time, the vendor.conf does have v0.4.3 properly.
